### PR TITLE
useToolsPanel: calculate menuItems in layout effect to avoid painting intermediate state

### PR DESCRIPTION
**Pull request by jsnajdr**
Fixes a visual glitch when displaying a tools panel in style editor. Go to global styles editor for individual blocks and in the list of blocks, click on Paragraph. At first, toolbar items with no content will appear:

tools-panel-empty

only after a moment the full UI is rendered:

tools-panel-final

The cause of this is how the state of the panel is updated, calculating derived state in a cascade of effects:
```js
const [ a, setA ] = useState();
const [ b, setB ] = useState();
const [ c, setC ] = useState();

useEffect( () => {
  setB( ... );
}, [ a ] );

useEffect( () => {
  setC( ... );
}, [ b ] );
```
This needs several rerenders and paints until the final state is achieved.

In our case, the state involved is the `menuItems` object and the `isShown` value derived from it.

This PR fixes that by using layout effect, so that all the updates happen without an intermediate paint that reveals the non-final UI state. See also #56536 which applied exactly the same fix at another place.

Only the first commit is needed to actually fix the glitch. The rest is using `useLayoutEffect` also for other state-updating effects in the same file. And removing some unneeded dependencies from `useCallback` and `useEffect` calls, namely state setters.

The next step could be a bigger refactoring to make the state updates more streamlined. We don't need effects: they only trigger rerenders. We could use `useMemo` or, for state updates that need to see the previous state, a `useReducer`.

### CODE REVIEW

**Comment by ciampo** [comment]
The changes make sense to me, although they seem to cause errors in unit tests. I guess we'd need to understand if we can improve the performance and timing of the updates, while keeping the expected behavior.

> Only the first commit is needed to actually fix the glitch. The rest is using `useLayoutEffect` also for other state-updating effects in the same file. And removing some unneeded dependencies from `useCallback` and `useEffect` calls, namely state setters.

Maybe removing the second commit would fix the unit tests?

**Comment by jsnajdr** [comment]
> they seem to cause errors in unit tests.

I fixed that by using layout effects also in `ToolsPanelItem`, so that the order of execution stays the same. The tool panels are a maze of fragile mutual updates, the code deserves a thorough review and refactor.

**Comment by ciampo** [comment]
> The tool panels are a maze of fragile mutual updates, the code deserves a thorough review and refactor.

Yeah, I got to the same conclusion while reviewing the PR. I think that moving the logic to a reducer could be a way to centralise state and logic a bit more.

**Review by ciampo — APPROVED** [review]
LGTM 🚀 

https://github.com/user-attachments/assets/6e875d75-ad25-4c83-9923-7da0547e5ff6


Notes / follow-ups:
- as mentioned earlier, `ToolsPanel` could benefit from an internal refactor. The logic is currently fragmented in a lot of places, making it fragile and hard to parse;
- I noticed a considerable delay when navigating to the block screen. We should look into how to take away that delay, either by improving loading performance of the screen, or by making it non-blocking (maybe show a skeleton UI while the real data is loading and a suspense boundary?)

**Comment by youknowriad** [comment]
Hello folks, this is causing a big error for me, when trying to edit an empty page in the site editor (just typing within an empty post content block). I think it's causing some kind of infinite loop. Reverting this PR solves the issue.

**Comment by youknowriad** [comment]
Also worth noting this PR caused an increase in "block select" metric numbers.

**Comment by jsnajdr** [comment]
There are infinite loops when the tools panel is used in combination with slots and fills. I tested only in Site Editor, in one specific panel that is not extensible. I missed the fatal errors this causes elsewhere.

> Also worth noting this PR caused an increase in "block select" metric numbers.

Interesting. That metric is measuring how long a `focus` or `focusin` event handlers take to execute. And this PR makes the tools panels render more "synchronously". Instead of rendering an empty placeholder initially and being populated with real content only after a cascade of `useEffect`s, the render is now immediate.

When preparing this PR for a second try, we should try to render the tools panel as a "low priority" update.

**Comment by stokesman** [comment]
> The next step could be a bigger refactoring to make the state updates more streamlined. We don't need effects:

Maybe we could take it one effect at a time. I happened to have #61694 which removes one and tests well for me.

**Comment by ciampo** [comment]
I personally like the approach in https://github.com/WordPress/gutenberg/pull/65564 as it's guaranteeing that all updates happen atomically, but happy to hear everyone's thoughts.

**Comment by stokesman** [comment]
> I personally like the approach in #65564

Me too. I hadn’t seen it when I made my prior comment and wish to add that the PR I mentioned is complementary and not an alternative.
